unread: Indicate which streams and topics have unread @-mentions.

jai2201 opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #21637 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

jai2201 commented:
Indicate Streams with unread-mention ->

<img width="275" alt="image" src="https://user-images.githubusercontent.com/76561593/183007978-de8cb50c-537b-4e25-ae67-efc7976e03bc.png">

timabbott commented on static/styles/app_components.css:
```diff
@@ -563,6 +564,10 @@ div.overlay {
 
+.unread_mention_info {
+    margin-right: 2px;
+}
```
Should this CSS rule apply to `:not(:empty)`, or do you want it to take up space even when empty?

jai2201 replied:
Hmmm, Sure!
Thanks for this!

timabbott commented on static/js/unread.js:
```diff
@@ -266,6 +267,14 @@ class UnreadTopicCounter {
+            const bucket = this.bucketer.reverse_lookup.get(message_id);
+            const stream_id = bucket.bucket_id;
+            if (!res.streams_with_mentions.includes(stream_id)) {
+                res.streams_with_mentions.push(stream_id);
```
Make this a `new Set`, and then you don't need to do this logic, which is quadratic.

jai2201 replied:
Updated the PR doing it. Thanks!

timabbott commented on static/js/unread.js:
```diff
@@ -574,6 +584,11 @@ export function num_unread_for_stream(stream_id) {
+export function stream_has_any_unread_mention_messages(stream_id) {
+    const res = unread_topic_counter.get_counts();
+    return res.streams_with_mentions.includes(stream_id);
+}
```
This seems like a bad API -- `get_counts()` is a somewhat expensive call getting global data; we definitely don't want to have a function like this that one might naturally call in a loop with that approach.

If you somehow need this API, it should just do the `const message_id of unread_mentions_counter` part of get_counts. But I suspect that in some places you're using it, the actually correct thing to do would be access the `streams_with_mentions` data in a `counts` object that you already got from `unread_topic_counter.get_counts`.

jai2201 replied:
Updated the PR, used `streams_with_mentions ` data from `counts`  object directly to get the answer for `stream_has_any_unread_mention_messages `

Thanks!

timabbott commented on static/js/stream_list.js:
```diff
@@ -348,6 +377,7 @@ export function update_dom_with_unread_counts(counts) {
     // counts.stream_count maps streams to counts
     for (const [stream_id, count] of counts.stream_count) {
         set_stream_unread_count(stream_id, count);
+        set_stream_unread_mention_info(stream_id);
```
This call, in particular, should access `counts.streams_with_mentions` directly, not call `get_counts` all over again.

jai2201 replied:
Implemented it in the style you asked, Thanks for teaching the way to avoid the expensive API call.

jai2201 commented:
I'm not sure why is this message arising even for a correct commit message format which is mentioned in our docs -- 
```
Use `Fixes part of #issue_number` for commits partly fixing an issue.
```

jai2201 commented on static/js/unread.js:
```diff
@@ -399,6 +399,27 @@ class UnreadTopicCounter {
+        }
+
+        const ids = Array.from(topic_bucket);
+
+        for (const id of ids) {
+            if (unread_mentions_counter.has(id)) {
+                return true;
+            }
+        }
```
@timabbott , I'm sure about is there any better approach through which we can get this thing done in less than `O(n)`.
Would like to see your suggestions, if any.

timabbott replied:
This is again an issue of having the wrong API. What you want to do is have a function `get_topics_with_unread_mentions(stream_id)` that returns a `Set` (and will be just as fast as this single-topic version), and call that outside the loop in `get_list_info`, and then inside the loop, you just need to check if the topic is inside the set, which is `O(1)`.

jai2201 replied:
Updated the PR removing this approach and tried to add a new one which does this operation in `O(1)`.
Thanks !

jai2201 commented:
Zoomed Out View for complete left-sidebar ->
<img width="291" alt="image" src="https://user-images.githubusercontent.com/76561593/183293186-864924dd-8bc6-4238-89e6-e4dcb444d9d2.png">

Zoomed In View for any stream in left-sidebar ->
<img width="291" alt="image" src="https://user-images.githubusercontent.com/76561593/183293224-c0810bf8-219a-4fbf-96af-12ffc836fe8a.png">

timabbott commented:
I did some rewriting of this, primarily inspired by the insight that the `reverse_lookup` function just needed a small tweak to return the `bucket_key` rather than the `bucket` to support this without larger changes to the `Bucketer` class. I also eliminated some code duplication in `stream_list.js`, and fixed case-sensitivity issues with topics and `undefined` being incorrectly added to the set of topics with unread mentions.

I'd appreciate your tweaking the tests so that they would fail if we didn't handle case-sensitivity correctly. 

There's a TODO I added that I'd like you to investigate.

jai2201 commented on static/js/stream_list.js:
```diff
@@ -335,6 +339,7 @@ class StreamSidebarRow {
 
     update_unread_count() {
         const count = unread.num_unread_for_stream(this.sub.stream_id);
+        // TODO: Do we need to check if there are unread mentions here?
         update_count_in_dom(this.$list_item, count);
     }
```
Thanks @timabbott  for the updates on PR; I went through the changes made and did the analysis on the TODO part you mentioned;

There's no need to update the part `stream_has_any_unread_mention_messages` here as even on doing it we get the value as `false` always initially and then I think the function to calculate `counts` is run and we get the value updated as `true` by running `set_stream_unread_count`;

jai2201 replied:
I would also like to propose to remove this whole thing - `udpate_unread_count()` from `StreamSidebar` class as it always runs the function `update_count_in_dom` with `count=0` only and then again the actual count gets updated by `set_stream_unread_count`, so what we can do is ->

remove this complete function from `StreamSidebar` and add `:not(:empty)` to `.unread_count` class in CSS;

I tried running the application with these updated changes and it looked fine;
let me know if there's anything I missed which makes this proposal wrong;

timabbott replied:
Yeah I think that's probably a fine approach, assuming you've confirmed that `count` is actually 0 -- how did you do that? I'd have expected it to depend on whether there are actually unreads if our initialization order is correct.

jai2201 replied:
I confirmed this by adding console statements; ->
Here's a screenshot which prints in order - `count`, `stream_id`, `place where console is written`;

<img width="480" alt="image" src="https://user-images.githubusercontent.com/76561593/185809325-1c4734d7-c79a-4256-9011-1aa72b953e80.png">


as you can see clearly that value of `count` is always 0 inside `update_unread_count()` and it gets updated inside `set_stream_unread_count()` which is called after this;

timabbott commented:
Also can you fix the node test? Might be best done in a new commit we plan to squash, just so it's easy for me to see your changes.

jai2201 commented:
@timabbott , I added a prep commit above and rebased the other commits accordingly and made each individual commit pass the CI, can you please take a look at them now?

timabbott commented:
Can you add a proper commit message to the new prep commit explaining why it's correct? Specifically, what code fills in the unread count already? And does the history suggest why this function existed if it's a noop? `git log -p` / `git log -S` may help explore that.

jai2201 commented:
Ahh, Thanks a lot @timabbott  for suggesting to look in the history why this function was introduced if it's a noop;

I looked through the commits in history and have found something tricky which makes this prep commit buggy;
Here's the link of the [commit](https://github.com/zulip/zulip/commit/c36932d29e540707a5fcab79600fb8d8c4bf366e) which introduces this piece of code.

jai2201 commented on static/js/stream_list.js:
```diff
@@ -335,7 +339,10 @@ class StreamSidebarRow {
 
     update_unread_count() {
         const count = unread.num_unread_for_stream(this.sub.stream_id);
-        update_count_in_dom(this.$list_item, count);
+        const stream_has_any_unread_mention_messages = unread.stream_has_any_unread_mention(
+            this.sub.stream_id,
+        );
+        update_count_in_dom(this.$list_item, count, stream_has_any_unread_mention_messages);
```
@timabbott  , I completed the TODO added by you and calculated the value of stream_has_any_unread_mention_messages here.

jai2201 commented on static/js/unread.js:
```diff
@@ -375,6 +375,21 @@ class UnreadTopicCounter {
         return util.sorted_ids(ids);
     }
 
+    get_streams_with_unread_mentions() {
+        const streams_with_mentions = new Set();
+        // Collect the set of streams containing at least one mention.
+        // We can do this efficiently, since unread_mentions_counter
+        // contains all unread message IDs, and we use stream_ids as
+        // bucket keys in our outer bucketer.
+
+        for (const message_id of unread_mentions_counter) {
+            const stream_id = this.bucketer.reverse_lookup.get(message_id);
+            streams_with_mentions.add(stream_id);
+        }
+
+        return streams_with_mentions;
+    }
+
```
I extracted this calculation in a separate function inside `UnreadTopicCounter` as I would not want to do complete calculation present inside `get_counts()` every time when I just need `streams_with_mentions`.

jai2201 commented on static/js/unread.js:
```diff
@@ -573,10 +618,19 @@ export function num_unread_for_topic(stream_id, topic_name) {
     return unread_topic_counter.get(stream_id, topic_name);
 }
 
+export function stream_has_any_unread_mentions(stream_id) {
+    const streams_with_mentions = unread_topic_counter.get_streams_with_unread_mentions();
+    return streams_with_mentions.has(stream_id);
+}
+
```
Added this separate function to just return the info about `unread_mentions` being present or not for a `stream_id` without much calculation;

jai2201 commented on static/js/stream_list.js:
```diff
@@ -335,7 +339,10 @@ class StreamSidebarRow {
 
     update_unread_count() {
         const count = unread.num_unread_for_stream(this.sub.stream_id);
-        update_count_in_dom(this.$list_item, count);
+        const stream_has_any_unread_mention_messages = unread.stream_has_any_unread_mentions(
+            this.sub.stream_id,
+        );
+        update_count_in_dom(this.$list_item, count, stream_has_any_unread_mention_messages);
```
Completed the TODO added by you here; and handled the case properly of pinning/unpinning any stream.

jai2201 commented:
@timabbott , I've updated the PR with completing the TODO added and made the node tests pass for each individual commit;

Left some comments above for you to have a look;
Would like to request for a re-test deployment of this PR on CZO.

timabbott commented on static/styles/app_components.css:
```diff
@@ -563,6 +563,11 @@ div.overlay {
     color: hsl(0, 0%, 100%);
 }
 
+.unread_mention_info:not(:empty) {
```
I think a 2px `margin-left` here would fix the overflow issue; test-deploying that but you might want to play with it.

jai2201 replied:
Cool, thanks @timabbott , I played with the PR and it looked fine with 2px margin-left;
pushed the changes;

timabbott commented on frontend_tests/node_tests/stream_list.js:
```diff
@@ -30,6 +30,7 @@ const scroll_util = mock_esm("../../static/js/scroll_util", {
 mock_esm("../../static/js/ui", {get_scroll_element: ($element) => $element});
 mock_esm("../../static/js/unread", {
     num_unread_for_stream: () => num_unread_for_stream,
+    stream_has_any_unread_mentions: () => stream_has_any_unread_mentions,
```
I'm going to merge this, but I think it'd be better to change this test to use the real `unread.js` module; it's a data module, so it should be easy to setup with a bit of data for the relevant streams rather than mocking it.

timabbott commented:
Merged, thanks for doing this @jai2201! Hopefully you've learned a bit about how to do simple but efficient data structures through this project :)
